src/box/txn_limbo.c

```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+			  "and empty limbo", reject_str(req),
+			  (long long)limbo->confirmed_lsn,
+			  (long long)lsn);
+		diag_set(ClientError, ER_SPLIT_BRAIN,
+			 "got a request mentioning future lsn");
+		return -1;
+	} else {
+		/*
+		 * Some entries are present in the limbo, we need to make sure
+		 * that request lsn lays inside limbo [first; last] range.
+		 * So that the request has some queued data to process,
+		 * otherwise it means the request comes from split brained node.
+		 */
+		int64_t first_lsn = txn_limbo_first_entry(limbo)->lsn;
+		int64_t last_lsn = txn_limbo_last_synchro_entry(limbo)->lsn;
+
+		if (lsn < first_lsn || last_lsn < lsn) {
+			say_error("%s. request lsn %lld out of range "
+				  "[%lld; %lld]", reject_str(req),
+				  (long long)lsn,
+				  (long long)first_lsn,
+				  (long long)last_lsn);
+			diag_set(ClientError, ER_SPLIT_BRAIN,
+				 "got a request lsn out of queue range");
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * Filter CONFIRM and ROLLBACK packets.
+ */
+static int
+filter_confirm_rollback(struct txn_limbo *limbo,
+			const struct synchro_request *req)
+{
+	assert(latch_is_locked(&limbo->promote_latch));
+	assert(limbo->validate);
+
+	/*
+	 * Zero LSN is allowed for PROMOTE and DEMOTE requests only.
+	 */
+	if (req->lsn == 0) {
+		say_error("%s. Zero lsn detected", reject_str(req));
+		diag_set(ClientError, ER_UNSUPPORTED, "Replication",
+			 "zero LSN for CONFIRM/ROLLBACK");
+		return -1;
+	}
+
+	return filter_queue_boundaries(limbo, req);
+}
+
+/** A filter PROMOTE and DEMOTE packets. */
+static int
+filter_promote_demote(struct txn_limbo *limbo,
+		      const struct synchro_request *req)
+{
+	assert(latch_is_locked(&limbo->promote_latch));
+	assert(limbo->validate);
+
+	/*
+	 * PROMOTE and DEMOTE packets must not have zero
+	 * term supplied, otherwise it is a broken packet.
+	 */
+	if (req->term == 0) {
+		say_error("%s. Zero term detected", reject_str(req));
+		diag_set(ClientError, ER_UNSUPPORTED,
+			 "Replication", "PROMOTE/DEMOTE with a zero term");
+		return -1;
+	}
+
+	/*
+	 * If the term is already seen it means it comes
+	 * from a node which didn't notice new elections,
+	 * thus been living in subdomain and its data is
+	 * no longer consistent.
+	 */
+	if (limbo->promote_greatest_term >= req->term) {
+		say_error("%s. Max term seen is %llu", reject_str(req),
+			  (long long)limbo->promote_greatest_term);
+		diag_set(ClientError, ER_SPLIT_BRAIN,
+			 "got a PROMOTE/DEMOTE with an obsolete term");
+		return -1;
+	}
+
+	return filter_queue_boundaries(limbo, req);
+}
+
+/** A fine-grained filter checking specific request type constraints. */
+static int
+txn_limbo_filter_request(struct txn_limbo *limbo,
+			 const struct synchro_request *req)
+{
+	if (!limbo->validate)
+		return 0;
+	/*
+	 * Wait until all the entries receive an lsn. The lsn will be
+	 * used to determine whether filtered request is safe to apply.
+	 */
+	if (txn_limbo_wait_persisted(limbo) < 0)
```

Comment:
That is a bit worrying. We do that in the most common case when we are the limbo owner and are writing normal transactions with correct CONFIRM. I suppose when you are the limbo owner and you write CONFIRM yourself, then you don't need to wait for last rows to finish WAL write, don't you? Could it be kept like that in some simple way?

Reply:
Oh, right, CONFIRM is written bypassing the txn_limbo_req flow (for now). Then should be fine atm.